File: its-my-training-day/leetcode-top-75/questions/arrays-hashing/greedy/best-time-buy-sell-stock/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

int bestTimeBuySellStock(const std::vector<int>& prices) {
    if (prices.empty()) return 0;
    int minPrice = prices[0];
    int maxProfit = 0;
    for (int i = 1; i < static_cast<int>(prices.size()); ++i) {
        if (prices[i] < minPrice) {
            minPrice = prices[i];
        } else {
            int profit = prices[i] - minPrice;
            if (profit > maxProfit) maxProfit = profit;
        }
    }
    return maxProfit;
}

File: its-my-training-day/leetcode-top-75/questions/arrays-hashing/greedy/best-time-buy-sell-stock/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const bestTimeBuySellStock = (prices: number[]): number => {
  if (prices.length === 0) return 0;
  let minPrice = prices[0];
  let maxProfit = 0;
  for (let i = 1; i < prices.length; i += 1) {
    if (prices[i] < minPrice) minPrice = prices[i];
    else {
      const profit = prices[i] - minPrice;
      if (profit > maxProfit) maxProfit = profit;
    }
  }
  return maxProfit;
};
